### ISSUE TITLE
refactor(ngControlName): exportAs ngControl instead of ngForm

### ISSUE THREAD

**Issue by PascalPrecht** [angular/angular#7236]
Something I'd like to discuss first.

It turns out, that `ngControl`/`NgControlName` is exported as `ngForm`. This is quite unintuitive when instances of `ngControl`are referenced. Say I want to display an error message when a control is invalid. I'd need an instance of that control to check it's validity:

```
<input ngControl="name" #name="ngForm">
<p *ngIf="!name.valid">This field is invalid!</p>
```

While this works (and I'm unsure if that's the way to do this), I'd rather expect `ngControl` to be exported as `ngControl`:

```
<input ngControl="name" #name="ngControl">
```

Is that something we can change? Or is there any reasoning behind `ngControl` being exported as `ngForm`?

This would be a BREAKING change.

**Comment by thelgevold** [comment]
Isn't this how to reference a field via the form
```
<form  [ngFormModel]="form" #f="ngForm">
    <div>
        <div class="formHeading">First Name</div>
        <input type="text" id="firstName" ngControl="firstName">
        <div class="errorMessage" *ng-if="f.form.controls.firstName.touched && !f.form.controls.firstName.valid">First Name is required</div>
    </div>
</form>
```

I don't follow why you're referencing the form on the input control. Maybe I am missing something?

**Comment by PascalPrecht** [comment]
@thelgevold Couple of things:

1) you're using `ngFormModel` which is model-driven forms. I'm currently exploring template-driven forms where you don't use that, as you don't have a control group created imperatively in your component

> I don't follow why you're referencing the form on the input control. Maybe I am missing something?

I'm not referencing the form. I'm referencing the `ngControl`, it just happens that it is exported as `ngForm` (just like `ngForm` is), which is exactly what I'm trying to point out here.

Basically, my goal was to get a reference of this particular control (as you can't access it via the control group with the `ngForm` instance), and it turned out it exports itself as `ngForm`. I expected `ngControl`.

**Comment by thelgevold** [comment]
@PascalPrecht Ah I see.

**Comment by PascalPrecht** [comment]
Hope that made it clear. If there's something I need to change in the issue description, please let me know.

**Comment by DanWahlin** [comment]
Definitely agree that exporting as ngForm isn't intuitive. Doesn't make sense at that level (input control level).

**Comment by laco0416** [comment]
BTW, I'm wondering, why is not it exported as `ngInput`?

**Comment by PascalPrecht** [comment]
@laco0416 that'd just yet another name to use. Except that `ngControl` prolly make more sense, because when you build custom controls, it doesn't necessarily have to be an input.

Still voting for `ngControl` and `ngControlGroup` respectively.

**Comment by laco0416** [comment]
@PascalPrecht Thanks! I agree to `ngControl`.

**Comment by thomashilzendegen** [comment]
+1

**Comment by SteffenJahr** [comment]
+1

**Comment by gingters** [comment]
+1

**Comment by PascalPrecht** [comment]
Can we please stop +1'ing this issue? There's a PR as base for a discussion, as soon as there's time things will move on.

**Comment by thomashilzendegen** [comment]
@PascalPrecht Yes, you're right. But in my opinion this breaking change, if everyone sees it in the same way as you, should be done soon, as this would affect many templates...

So this should just put some attention to this pending issue. :)

**Comment by PascalPrecht** [comment]
> So this should just put some attention to this pending issue. :)

This just puts attention to our always ringing inbox as comments are made. It's just annoying and doesn't cause any speed-up.

Patience please.

**Comment by wardbell** [comment]
I'm going to disagree with you @PascalPrecht.  Well ... not so much disagree as justify the choice of "ngForm".

For reader's who don't know,  a directive's `exportAs` property tells Angular how to link local variable to the directive as in 
```
<input ngControl="name" #name="ngForm">
```
As Pascal observes, we set the `name` local variable to `ngForm` because the `NgControlName` directive's [exportAs](https://angular.io/docs/ts/latest/api/core/DirectiveMetadata-class.html#!#exportAs) property happens to be "ngForm".
    
This seems unintuitive because `name` becomes a reference to an instance of the `NgControlName` directive. Why not call it "NgControl"?

Further investigation reveals that *all* control directives in the  Angular form family &mdash; including `NgForm`, `NgModel`, `NgControlName` and `NgControlGroup` &mdash; *exportAs* "ngForm".

Rather than give each its own *exportAs* value (which we'd have to guess), they all get the same one, the name of their directive *family*, "ngForm".

Consistency rules!

For me, once explained, it makes perfect sense.
We only ever apply *one* of these directives to an element tag so there's never a potential conflict.

**Comment by frederikschubert** [comment]
I think @wardbell 's explanation makes sense but the wording is not intuitive. I would rather have the form directives export themselves as `ngFormDirective`. As previously stated there will always be only one directive from the ngForm family on an element.
This way we have consistency and also have a more intuitive binding e.g. `<input ngControl="username" #username="ngFormDirective">`.

**Comment by wardbell** [comment]
I think we DO need to talk about the "Forms Family Of Directives (tm)".  I updated the Forms chapter and took a small swing at this ... and clarifying the use of `ngForm` in setting the local var.

I want also to explain a bit more about how local vars and `exportAs` work in the Template-Syntax chapter.

**Comment by cagataycivici** [comment]
I agree, I read the forms docs and got super confused because of this naming convention. ngControl="name" #name="ngControl" would have been a better choice.

**Comment by zoechi** [comment]
Actually I also don't see any value in consistency. Intention revealing is key for me in naming.

**Comment by victorcarvalhosp** [comment]
+1

**Comment by CaptainCodeman** [comment]
> Rather than give each its own exportAs value (which we'd have to guess)

Why would we have to guess? `<input ngControl="name" #name="ngControl">` seems much clearer to me. Using `ngForm`

**Comment by jjk3yz** [comment]
Having all the form directives exported as 'ngForm' is very confusing and leads to a poor developer experience. Name's exist for a reason and the art of naming things appropriately helps us to make sense of the world. It's noteworthy to add that it's much more important to get it right when dealing with abstract things i.e software abstractions. 

So lets call an ngForm an ngForm and an ngControl an ngControl. Yes it's a breaking change but it's a sorely needed one.

**Comment by xe4me** [comment]
I've read the whole form section 4 times , ngFormModel ngFrom ngControl ngFormControl , the most obscure naming convention ever.

**Comment by sam-s4s** [comment]
Yes, the fact that NgForm, NgModel, NgControlName and NgControlGroup all export as ngForm is quite confusing, and feels cluttered.

**Comment by kara** [comment]
Closing this, as the new forms module replaces the `ngForm` export names with names matching the directive:

`ngModel` exports as "ngModel" (as of 0.1.1 and RC.3)
`ngModelGroup` exports as "ngModelGroup"

**Comment by DanWahlin** [comment]
Great job on this Kara - it's how it should be now. :-)

**Comment by niwsa** [comment]
I know this is not the right place to post something about the docs or guide but can someone attest to the fact that the doc at this [https://angular.io/docs/ts/latest/guide/forms-deprecated.html#!#ngForm](url) link is using #name="ngForm" instead of #name="ngModel".Quite often I jump into some inconsistencies/incomplete info in the doc/guide/tutorial which leaves me guessing. :)

**Comment by kara** [comment]
@niwsa If you look at the top of the page you linked, you'll see that it says it's the guide to the "deprecated forms API" :)  So it's actually correct in suggesting the ngForm export for that old module.  It also links to the new forms docs, which are here: https://angular.io/docs/ts/latest/guide/forms.html. They are properly updated to use the ngModel export.

Soon we'll be switching over to have the new forms docs as the default on display to avoid this type of confusion. Hopefully you didn't lose too much time!

**Comment by niwsa** [comment]
Thanks @kara .I think I should exercise some restraint before posting something looks like I have been spamming some of these forums for a while.:P

**Comment by kara** [comment]
No worries; glad i could point you to the updated docs. There are more improvements coming!
